### PR TITLE
Fixes for bonnie integration

### PR DESCRIPTION
**Pull request by daco101**
This PR contains a collection of fixes and code ports from bonnie that were created while working on integrating this into bonnie.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/BONNIE-1916
- [x] Internal ticket links to this PR
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass

**Reviewer 1:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

**Reviewer 2:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

File: lib/measure-loader/cql_loader.rb [hunk 1]
```diff
@@ -20,6 +20,8 @@ def extract_measures
       end
       measure.package = CQM::MeasurePackage.new(file: BSON::Binary.new(@measure_zip.read))
       measures << measure
+
+      update_population_set_and_strat_titles(measures, @measure_details[:population_titles])
```
Comment (losborne) [inline lib/measure-loader/cql_loader.rb]: These helper functions are great.  A small improvement in readability I would make is to define them below in the order they are used (`update_population_set_and_strat_titles`, then `generate_vs_items_and_verify_valuesets_complete`, then `verify_hqmf_valuesets_match_elm_valuesets`).

File: lib/measure-loader/cql_loader.rb [hunk 2]
```diff
@@ -52,25 +55,30 @@ def create_measure(measure_files)
...
-      measure.cql_libraries = cql_libraries
       vs_items[:value_set_models].each { |vsm| measure.value_sets.push vsm }
+      measure.cql_libraries = cql_libraries
       measure.composite = measure_files.components.present?
 
       return measure
     end
 
+    def verify_hqmf_valuesets_match_elm_valuesets(elm_valuesets, measure_hqmf_model)
+      # Exclude patient birthdate and expired OIDs used by SimpleXML parser for AGE_AT handling and bad oid protection in missing VS check
```
Comment (losborne) [inline lib/measure-loader/cql_loader.rb]: Could you expand this comment a bit to clarify which oids below refer to which cases mentioned in the comment?

File: lib/measure-loader/cql_loader.rb [hunk 3]
```diff
@@ -87,6 +95,14 @@ def create_cql_libraries(cql_library_files, main_cql_lib)
       return cql_libraries
     end
 
+    def generate_vs_items_and_verify_valuesets_complete(cql_libraries, hqmf_model)
```
Comment (losborne) [inline lib/measure-loader/cql_loader.rb]: I think this should be 3 separate small functions: `generate_elm_valuesets`, `verify_hqmf_valuesets_match_elm_valueset` (no change), and `generate_vs_items`.  Except for the last function, I think the term items `vs_items` could be swapped for something more descriptive.  Are these just `value_sets`? or what distinguishes them?

File: lib/measure-loader/hqmf_measure_loader.rb [hunk 4]
```diff
@@ -43,11 +44,27 @@ def create_measure_model(hqmf_xml, hqmf_model_hash, measure_scoring)
...
+
       def extract_population_set_models(hqmf_xml, measure_scoring)
         populations = hqmf_xml.css('/QualityMeasureDocument/component/populationCriteriaSection')
-        return populations.map do |population|
+        return populations.map.with_index do |population, pop_index|
           ps_hash = extract_population_set(population)
-          population_set = CQM::PopulationSet.new(title: ps_hash[:title], population_set_id: ps_hash[:id])
+          population_set = CQM::PopulationSet.new(
+            title: ps_hash[:title],
+            population_set_id: "PopulationSet_#{pop_index+1}"
```
Comment (losborne) [inline lib/measure-loader/hqmf_measure_loader.rb]: Why the `+1` here and on lines 80/81 below?
```
              stratification_id: "#{population_set.population_set_id}_Stratification_#{index+1}",
              title: "PopSet#{pop_index+1} Stratification #{index+1}",
```

It seems off to have that mismatched with `pop_index` or `index` to me.

File: lib/measure-loader/hqmf_measure_loader.rb [hunk 5]
```diff
@@ -89,11 +106,15 @@ def extract_population_set(population_hqmf_node)
...
             ps[:populations][HQMF::PopulationCriteria::DENEX] = statement_ref_string
+          when 'denominatorExceptionCriteria'
+            ps[:populations][HQMF::PopulationCriteria::DENEXCEP] = statement_ref_string
           when 'measurePopulationCriteria'
             ps[:populations][HQMF::PopulationCriteria::MSRPOPL] = statement_ref_string
           when 'measurePopulationExclusionCriteria'
             ps[:populations][HQMF::PopulationCriteria::MSRPOPLEX] = statement_ref_string
           when 'stratifierCriteria'
+            # Ignore Supplemental Data Elements
+            next if cc.at_css('component[@typeCode="COMP"]/measureAttribute/code[@code="SDE"]').present?
```
Comment (losborne) [inline lib/measure-loader/hqmf_measure_loader.rb]: Recommend moving `cc.at_css('component[@typeCode="COMP"]/measureAttribute/code[@code="SDE"]').present` to a function with name something like `has_supplemental_data_elements`

File: lib/measure-loader/vsac_value_set_loader.rb [hunk 6]
```diff
@@ -45,6 +46,20 @@ def retrieve_and_modelize_value_sets_from_vsac(value_sets, measure_id = nil)
 
     private
 
+    def vs_api()
```
Comment (losborne) [inline lib/measure-loader/vsac_value_set_loader.rb]: Recommend naming this `get_vs_api` or something else with a verb--reading `vs_api.get_multiple_valuesets(needed_value_sets)`, it seemed like `vs_api` was an undefined variable to me.

File: lib/measure-loader/vsac_value_set_loader.rb [hunk 7]
```diff
@@ -31,7 +32,7 @@ def retrieve_and_modelize_value_sets_from_vsac(value_sets, measure_id = nil)
         end
       end
 
-      vs_responses = @api.get_multiple_valuesets(needed_value_sets)
+      vs_responses = vs_api.get_multiple_valuesets(needed_value_sets)
```
Comment (losborne) [inline lib/measure-loader/vsac_value_set_loader.rb]: Not sure if rubocop would agree or not, but I think adding the () after the initial function call would make it clearer that this is calling a function rather than referring to a variable (see comment below about name as well).

File: lib/measure-loader/vsac_value_set_loader.rb [hunk 8]
```diff
@@ -45,6 +46,20 @@ def retrieve_and_modelize_value_sets_from_vsac(value_sets, measure_id = nil)
 
     private
 
+    def vs_api()
+      return @api if @api.present?
+      @api = Util::VSAC::VSACAPI.new(config: APP_CONFIG['vsac'], ticket_granting_ticket: @vsac_ticket_granting_ticket, username: @vsac_username, password: @vsac_password)
+      return @api
+    end
+
+    def get_ticket_granting_ticket(vsac_username, vsac_password)
```
Comment (losborne) [inline lib/measure-loader/vsac_value_set_loader.rb]: It looks like `def get_ticket_granting_ticket` now exists 3 times in this repo--should the old one be removed?

File: test/fixtures/vcr_cassettes/measure__test_cms_id.yml [hunk 9]
```diff
@@ -0,0 +1,4739 @@
...
+  recorded_at: Wed, 06 Feb 2019 17:49:43 GMT
+- request:
+    method: get
+    uri: https://vsac.nlm.nih.gov/vsac/svs/RetrieveMultipleValueSets?id=2.16.840.1.113883.3.464.1003.101.12.1001&profile=MU2%20Update%202016-04-01&ticket=ST-7307201-nfFPmFsyDaEUBJejzpQr-cas
+    body:
+      encoding: US-ASCII
+      string: ''
+    headers:
+      User-Agent:
+      - Typhoeus - https://github.com/t
```
Comment (losborne) [inline test/fixtures/vcr_cassettes/measure__test_cms_id.yml]: It's weird that this one is `N08` and not `N08.8` like the others o.O

File: lib/hqmf-parser/cql/document_helpers/doc_population_helper.rb [hunk 10]
```diff
@@ -43,7 +43,11 @@ def extract_observations
           # The criteria_reference_id is the id of the measurePopulationCriteria that should be used for this observation function
           measure_population_id = entry.at_xpath("*/cda:measureObservationDefinition/cda:component/cda:criteriaReference/cda:id").attributes['root'].value
           # Get the name of the parameter to the  observation function within the measurePopulationCriteria section
-          cql_define_function[:parameter] = @doc.at_xpath("cda:QualityMeasureDocument/cda:component/cda:populationCriteriaSection/cda:component/cda:measurePopulationCriteria/cda:id[@root = \"#{measure_population_id}\"]/../cda:precondition/cda:criteriaReference/cda:id").attributes['extension'].value.match('\\"([A-Za-z0-9 ]+)\\"')[1]
+
+          criteriaReferenceId = @doc.at_xpath("cda:QualityMeasureDocument/cda:component/cda:populationCriteriaSection/cda:component/cda:measurePopulationCriteria/cda:id[@root = \"#{measure_population_id}\"]/../cda:precondition/cda:criteriaReference/cda:id")
+          #alternate path? why?
```
[the remark this answers is not in the document]
Comment (daco101) [reply lib/hqmf-parser/cql/document_helpers/doc_population_helper.rb]: Still unknown

File: lib/measure-loader/hqmf_measure_loader.rb (hunk 4 above)
Comment (daco101) [reply lib/measure-loader/hqmf_measure_loader.rb]: I think it's more friendly to users this way.

Comment (losborne) [reply lib/measure-loader/hqmf_measure_loader.rb]: Ok, since it's in a String, that makes sense to me.

File: test/fixtures/vcr_cassettes/measure__test_cms_id.yml (hunk 9 above)
Comment (daco101) [reply test/fixtures/vcr_cassettes/measure__test_cms_id.yml]: What do you mean?

Comment (losborne) [reply test/fixtures/vcr_cassettes/measure__test_cms_id.yml]: The line I commented on changed or something--It's not important, just noticed that one of the elements in the .yml file didn't match the others--nothing to worry about.

File: lib/measure-loader/cql_loader.rb [hunk 11]
```diff
@@ -20,22 +20,62 @@ def extract_measures
...
+        strat.title = population_titles[title_idx] if population_titles[title_idx].present?
+        title_idx += 1
+        break if title_idx >= population_titles.size
+      end
+    end
+
     private
 
     def create_measure_and_components(measure_files)
+      top_level_library_ids = measure_files.cql_libraries.map { |lib| "#{lib.id}_v#{lib.version}" }
```
Comment (jbradl11) [inline lib/measure-loader/cql_loader.rb]: Is "top_level_library" what we are calling composite measure support libraries?

Comment (daco101) [reply lib/measure-loader/cql_loader.rb]: Yah it's what I was going to call anything in the top composite measure folder - so that's the support libraries and the main composite library. But there's been some push back on the name, so feel free to weigh in with a suggestion. Zach suggests "top_level_composite_library". The real use of the field is to decide whether to show the function definition in the bonnie front-end / excel, but since this is being set in cqm-parsers I can't use a bonnie specific name.

Comment (daco101) [reply lib/measure-loader/cql_loader.rb]: We have resolved to leave this as is.
